### PR TITLE
Add new excludeRepos option to the Gitlab provider

### LINKED ISSUE #25405: 🚀 Feature: Allow repositories to be excluded from discovery
### 🔖 Feature description

Currently the discovery integrations (at least, the ones I'm using - Azure and Gitlab) have options to set what repositories should be included, e.g. `projectPattern` for the [Gitlab discovery plugin](https://backstage.io/docs/integrations/gitlab/discovery), however there is no option to exclude a repository from discovery.

I would like a way to provide a list of repositories to the discovery plugins which should not be included during discovery. For example, a new configuration option like `excludeRepositories` which takes a list of project paths:

```yaml
  providers:
    gitlab:
      our_org_gitlab:
        host: gitlab.myorg.co.uk
        branch: main
        fallbackBranch: master
        excludeRepositories:
          - foo/bar
          - foo/qux
          - group/subgroup/repo
```

A similar option for the Azure DevOps plugin could just take a list of repository names.

Repositories marked for exclusion would then not be discovered.

For me, this specifically relates to the Gitlab discovery plugin, but I think it would be useful on the others as well.

### 🎤 Context

The obvious solution here is to simply not add a `catalog-info.yaml` file to the repository to exclude it from discovery, right?

However, we currently have the situation where some of our repositories are hosted in Azure DevOps, and the code is then synced to our Gitlab instance where our CI/CD pipelines execute. We have a large number of repositories configured like this, but we also have an even larger number of projects that only exist within the Gitlab instance.

We would like to run discovery on both Azure DevOps and Gitlab, however the synced projects will be discovered twice, in two locations.

Ideally we would be able to exclude the synced repositories from discovery in one of the providers, so that we can avoid them being detected twice.

Using the include functionality (i.e. `projectPattern`) won't work for us since the project names and paths are many and varied, and it would require a large, complex, negative-lookahead regex to even start to make this work.

### ✌️ Possible Implementation

N/A

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Xiol**
## Hey, I just made a Pull Request!

Adds a new `excludeRepos` option to the Gitlab catalog provider to resolve #25405.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend-module-gitlab/src/lib/types.ts [hunk 1]
```diff
@@ -201,6 +201,11 @@ export type GitlabProviderConfig = {
    * If the project is a fork, skip repository
    */
   skipForkedRepos?: boolean;
+  /**
+   * List of repositories to exclude from discovery, should be the full path to the repository, e.g. `group/project`
+   * Paths should not start or end with a slash.
+   */
+  excludeRepos: string[];
```
Comment (benjdlambert) [inline plugins/catalog-backend-module-gitlab/src/lib/types.ts:208]: Let's make this optional just to be safe :pray:

File: .changeset/red-radios-promise.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-backend-module-gitlab': minor
```
Comment (benjdlambert) [inline .changeset/red-radios-promise.md:2]:
```suggestion
'@backstage/plugin-catalog-backend-module-gitlab': patch
```

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Hey :wave: thanks for this, just a few bits :pray:

File: plugins/catalog-backend-module-gitlab/src/lib/types.ts (hunk 1 above)
Comment (Xiol) [reply plugins/catalog-backend-module-gitlab/src/lib/types.ts]: Sorry, I'm still feeling my way around TypeScript. I'm assuming I just need the `?` suffix to the property name, which I've now added. Thanks

**Review by benjdlambert — APPROVED** [review]
Nice! Thanks for this :pray:

**Comment by benjdlambert** [comment]
Ah, just though of one thing - you might want to update the `config.d.ts` schema too to add this additional field to the config options with a nice description :pray:

https://github.com/backstage/backstage/blob/b357c6166bbb9aa1dcc92f7ca8a7fa397aff3f97/plugins/catalog-backend-module-gitlab/config.d.ts

**Pull request merged**